File: pa4/List.c
// Reginald Chand
// rzchand
// CMPS 101
// Doubly linked list for sorting a List alphabetically
// List.c

#include<stdio.h>
#include<stdlib.h>
#include<assert.h>
#include"List.h"

// private types and functions ------------------------------------------------
// NodeObj
typedef struct NodeObj{
    int* item;
    struct NodeObj* next;
    struct NodeObj* prev;
} NodeObj;

// NodeObj
typedef NodeObj* Node;

// newNode()
// constructor for private Node typedef
Node newNode(int* x){
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N->item = x;
	N->next = NULL;
	N->prev = NULL;
	return N;
}

// freeNode()
// destructor for private Node type
void freeNode(Node* pN){
	if(pN != NULL && *pN != NULL){
		free(*pN);
		*pN = NULL;
	}
}

// ListObj
typedef struct ListObj{
	Node front;
	Node back;
	Node cursor;
	int numItems;
	int cursIndx;
} ListObj;

// deletes all Nodes
void deleteAll(List L){
	Node A = NULL;
	while(L->front != NULL){
		A = L->front;
		L->front = L->front->next;
		freeNode(&A);
	}
}

// returns 0 if list is empty
int isEmpty(List L){
	return (L->numItems == 0);
}
// public functions -----------------------------------------------------------
List newList(void){
	List L = malloc(sizeof(ListObj));
	assert(L != NULL);
	L->front = NULL;
	L->back = NULL;
	L->cursor = NULL;
	L->cursIndx = -1;
	L->numItems = 0;
	
	return L;
}

void freeList(List* pL){
	assert(pL !=NULL && *pL != NULL);
	while(!isEmpty(*pL)){
		deleteBack(*pL);
	}	
	free(*pL);
	*pL = NULL;
	
}

int length(List L){
	return L->numItems;
}

int index(List L){
	if(L->cursor == NULL){
		return -1;
	}
	return L->cursIndx;
}

int front(List L){
	if(L->front == NULL){
		fprintf(stderr,"List ADT Error: calling front() on empty list");
	}
	
	return L->front->item;
}

int back(List L){
	if(L->back == NULL){
		fprintf(stderr,"List ADT Error: calling back() on empty list");
	}
	
	return L->back->item;
}

int get(List L){
	if(L->cursIndx < 0){
		fprintf(stderr,"List ADT Error: calling get() while cursor is undefined");
	}
	if(L == NULL){
		fprintf(stderr,"List ADT Error: calling get() on an empty list");
	}
	
	return L->cursor->item;
}

int equals(List A, List B){
	if(A == NULL || B == NULL){
		fprintf(stderr,"List ADT Error: calling equals() with an empty list");
	}
	
	Node temp = A->front;
	Node temp2 = B->front;
	
	while(temp != NULL && temp2 != NULL){
		// if sequence does not match, break from loop
		if(temp->item != temp2->item){
			break;
		}
		
		temp = temp->next;
		temp2 = temp2->next;
	}
	// if the while loop completes
	if(temp == temp2){
		return 1;
	}
	else{
		return 0;
	}
}

void clear(List L){
	deleteAll(L);
	L->front = NULL;
	L->back = NULL;
	L->cursor = NULL;
	L->cursIndx = -1;
	L->numItems = 0;
}

void moveFront(List L){
	if(!isEmpty(L)){
		L->cursor = L->front;
		L->cursIndx = 0;
	}
}

void moveBack(List L){
	if(!isEmpty(L)){
		L->cursor = L->back;
		L->cursIndx = L->numItems - 1;
	}
}

void movePrev(List L){
	// if cursor == front
	if(L->cursIndx == 0 && L->cursor != NULL){
		L->cursor = NULL;
		L->cursIndx = -1;
	}
	// if cursor > front
	else if(L->cursIndx > 0 && L->cursor != NULL){
		L->cursor = L->cursor->prev;
		L->cursIndx--;
	}
}

void moveNext(List L){
	// if cursor == back
	if(L->cursIndx == L->numItems - 1){
		L->cursor = NULL;
		L->cursIndx = -1;
	}
	// if cursor < back
	else if(L->cursIndx < L->numItems - 1){
		L->cursor = L->cursor->next;
		L->cursIndx++;
	}
}

void prepend(List L, int data){
	if(L == NULL){
		fprintf(stderr,"List ADT Error: calling prepend() with an empty list");
	}
	
	Node N = newNode(data);
	if(isEmpty(L)){
		L->front = N;
		L->back = L->front;
	}
	else{
		N->next = L->front;
		L->front->prev = N;
		L->front = N;
		
		// checks if cursor index needs to be adjusted
		if(L->cursIndx != -1){
			L->cursIndx++;
		}
	}
	L->numItems++;
}
void append(List L, int data){
	Node N = newNode(data);
	if(isEmpty(L)){
		L->front = N;
		L->back = N;
	}
	else{
		N->prev = L->back;
		L->back->next = N;
		L->back = N;
	}
	L->numItems++;
}
void insertBefore(List L, int data){
	if(L == NULL || isEmpty(L)){
		fprintf(stderr,"List ADT Error: calling insertBefore() with an empty list");
	}
	if(L->cursIndx < 0 || L->cursor == NULL){
		fprintf(stderr,"List ADT Error: calling insertBefore() with undefined cursor");
	}
	Node N = newNode(data);
	if(L->cursor == L->front){
		N->next = L->front;
		L->front->prev = N;
		L->front = N;
	}
	else{
		N->prev = L->cursor->prev;
		L->cursor->prev->next = N;
		N->next = L->cursor;
		L->cursor->prev = N;
	}
	L->numItems++;
}
void insertAfter(List L, int data){
	if(L == NULL || isEmpty(L)){
			fprintf(stderr,"List ADT Error: calling insertAfter() on an empty list");
		}
		if(L->cursIndx < 0 || L->cursor == NULL){
			fprintf(stderr,"List ADT Error: calling insertAfter() with undefined cursor");
		}
		Node N = newNode(data);
		if(L->numItems > 1){
			N->prev = L->cursor;
			N->next = L->cursor->next;
			
			L->cursor->next->prev = N;
			L->cursor->next = N;
		}
		else{
			N->prev = L->front;
			L->front->next = N;
		}
	L->numItems++;
}	

void deleteFront(List L){
	if(L == NULL){
		fprintf(stderr,"List ADT Error: calling deleteFront() on NULL list");
	}
	if(L->numItems < 1){
		fprintf(stderr,"List ADT Error: calling deleteFront() on an empty list");
	}
	// check if cursor is at the front,
	// set cursor to NULL if it is
	if(L->cursIndx == 0){
		L->cursIndx = -1;
		L->cursor = NULL;
	}
	
	Node temp = L->front;
	L->front = L->front->next;
	L->front->prev = NULL;
	L->numItems--;
	freeNode(&temp);
}
void deleteBack(List L){
	if(L == NULL){
		fprintf(stderr,"List ADT Error: calling deleteBack() on an NULL list");
	}
	if(L->numItems < 1){
		fprintf(stderr,"List ADT Error: calling deleteFront() on an empty list");
	}
	// check if cursor was at the back
	if(L->cursIndx == L->numItems - 1){
		L->cursIndx = -1;
		L->cursor = NULL;
	}
	Node temp = L->back;
	if(L->numItems > 1){
		L->back = L->back->prev;
		L->back->next = NULL;

	}
	else{
		L->front = L->back = L->cursor = NULL;
	}
	L->numItems--;
	freeNode(&temp);
}
void delete(List L){
	if(L == NULL){
		fprintf(stderr,"List ADT Error: calling delete() on an empty list");
	}
	if(L->cursIndx < 0){
		fprintf(stderr,"List ADT Error: calling delete() when cursor is undefined");
	}
	
	Node temp = L->cursor;
		
	if(L->cursIndx == 0){
		deleteFront(L);
	}
	else if(L->cursIndx == L->numItems - 1){
		deleteBack(L);
	}
	else{
		L->cursor->prev->next = L->cursor->next;
		L->cursor->next->prev = L->cursor->prev;
		L->cursor->next = L->cursor->prev = NULL;
		
		L->numItems--;
		
		freeNode(&temp);
		L->cursor = NULL;
		L->cursIndx = -1;	
	}	
}

void printList(FILE* out, List L){
   if( L == NULL ){
      fprintf(stderr, 
         "Dictionary Error: calling printDictionary() on NULL"
         " Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   
   for(Node N = L->front; N != NULL; N = N->next){
	   fprintf(out, "%d ", N->item);
   }
}

List copyList(List L){
	List listCpy = newList();
	Node temp = L->front;
	
	while(temp != NULL){
		append(listCpy, temp->item);
		temp = temp->next;
	}
	return listCpy;
}

File: pa3/List.java
//# Reginald Chand
//# rzchand
//# CMPS 101

// List.java
// A bi-directional queue List ADT

public class List{
	private class Node{
		Object item;
		Node next;
		Node prev;
		
		Node(Object x){
			item = x;
			next = null;
			prev = null;
		}
		
		public boolean equals(Object x){
		   boolean eq = false;
		   Node N;
		   
		   if(x instanceof Node){
			  N = (Node)x;
			  eq = this.item.equals(N.item);
		   }
		   
		   return eq;
		}
	}

	private Node front; // reference to the first Node in the List
	private Node back; // reference to the last Node in the List
	private Node cursor; 
	private int numItems;
	private int cursIndx;

	private boolean isEmpty(){
		return numItems == 0;
	}
	// constructor for List class
	List(){
		front = null;
		back = null;
		cursor = null;
		cursIndx = -1;
		numItems = 0;
	}
	
	// Access functions -------------------------------
	// Returns the number of elements in this List.
	int length(){
		return numItems;
	}
	// If cursor is defined, returns the index of the cursor element,
	// otherwise returns -1.
	int index(){
		if(cursor == null){
			return -1;
		}
		return cursIndx;
	}
	// Returns front element. Pre: length()>0
	Object front(){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling front() on empty list");
		}

		return front.item;
	
	}
	// Returns back element. Pre: length()>0
	Object back(){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling front() on empty list");
		}
		
		return back.item;
	}
	// Returns cursor element. Pre: length()>0, index()>=0
	Object get(){
		if(cursIndx < 0){
			throw new RuntimeException("List ADT Error: calling get() while cursor is undefined");
		}
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling get() on an empty list");
		}
		return cursor.item;
	}

	// Returns true if this List and L are the same integer
	// sequence. The cursor is ignored in both lists.
	public boolean equals(List L){
		if(L == null){
			throw new RuntimeException("List ADT Error: List is undefined");
		}
	
		Node temp = front;
		Node temp2 = L.front;
		
		while(temp != null && temp2 != null){
			// if the sequence does not match it'll return false
			if(!temp.equals(temp2)){
				return false;
			}
			
			temp = temp.next;
			temp2 = temp2.next;
		}
		return true;
	}
	
	// Manipulation procedures
	// Resets this List to its original empty state.
	void clear(){
		if(this != null){
			front = null;
			back = null;
			cursor = null;
			cursIndx = -1;
			numItems = 0;
		}
	}
	// If List is non-empty, places the cursor under the front element,
	// otherwise does nothing.
	void moveFront(){
		if(!(this.isEmpty())){
			cursor = front;
			cursIndx = 0;
		}
	}
	// If List is non-empty, places the cursor under the back element,
	// otherwise does nothing.
	public void moveBack(){
		if(!(this.isEmpty())){
			cursor = back;
			cursIndx = numItems - 1;
		}
	}
	// If cursor is defined and not at front, moves cursor one step toward
	// front of this List, if cursor is defined and at front, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void movePrev(){
		// if cursor == front
		if(cursIndx == 0 && cursor != null){
			cursor = null;
			cursIndx = -1;
		}
		// if cursor > front
		else if(cursIndx > 0 && cursor != null){
			cursor = cursor.prev;
			cursIndx--;
		}
	}
	// If cursor is defined and not at back, moves cursor one step toward 
	// back of this List, if cursor is defined and at back, cursor becomes
	// undefined, if cursor is undefined does nothing.
	void moveNext(){
	  if(cursor != null && cursIndx != numItems - 1) {
         cursor = cursor.next;
         cursIndx++;
      }
      else if(cursor != null && cursIndx == numItems - 1) {
         cursor = null;
         cursIndx = -1;
      }
	}
	// Insert new element into this List. If List is non-empty,
	// insertion takes place before front element.
	void prepend(Object data){
		if(this.isEmpty()){
			front = new Node(data);
			back = front;
			numItems++;
		}
		else{
			Node temp = new Node(data);
			temp.next = front;
			front.prev = temp;
			front = temp;
			numItems++;
			// checks if cursor index needs to be adjusted after adding new node
			if(cursIndx != -1){
				cursIndx++;
			}
		}
	}
	// Insert new element into this List. If List is non-empty,
	// insertion takes place after back element.
	void append(Object data){
		if(this.isEmpty()){
			front = new Node(data);
			back = front;
			numItems++;
		}
		else{
			Node temp = new Node(data);
			temp.prev = back;
			back.next = temp;
			back = temp;
			numItems++;
		}
	}
	// Insert new element before cursor.
	// Pre: length()>0, index()>=0
	void insertBefore(Object data){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling insertBefore on an empty list");
		}
		if(cursIndx < 0 || cursor == null){
			throw new RuntimeException("List ADT Error: calling insertBefore() with undefined cursor");
		}
		if(numItems > 1 && cursIndx > 0){
			Node temp = new Node(data);
			temp.next = cursor;
			temp.prev = cursor.prev;
			
			cursor.prev = temp;
			temp.prev.next = temp;
		}
		else if(numItems > 1 && cursIndx == 0){
			Node temp = new Node(data);
			temp.next = front;
			front.prev = temp;
			front = temp;
		}
		else{
			Node temp = new Node(data);
			temp.next = front;
			front.prev = temp;
			front = temp;
		}
		
		numItems++;
		// cursor index need to be adjusted
		cursIndx++;
	}
	// Inserts new element after cursor.
	// Pre: length()>0, index()>=0
	void insertAfter(Object data){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling insertAfter() on an empty list");
		}
		if(cursIndx < 0 || cursor == null){
			throw new RuntimeException("List ADT Error: calling insertAfter() with undefined cursor");
		}
		if(numItems > 1){
			Node temp = new Node(data);
			temp.prev = cursor;
			temp.next = cursor.next;
			
			cursor.next.prev = temp;
			cursor.next = temp;
		}
		else{
			Node temp = new Node(data);
			temp.prev = front;
			front.next = temp;
		}
		numItems++;
	}
	// Deletes the front element. Pre: length()>0
	void deleteFront(){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling deleteFront() on an empty list");
		}
		// check if cursor is at the front,
		// set cursor to null if it is
		if(cursIndx == 0){
			cursIndx = -1;
			cursor = null;
		}
		front = front.next;
		front.prev = null;
		numItems--;
	}
	// Deletes the back element. Pre: length()>0
	void deleteBack(){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling deleteBack() on an empty list");
		}
		// check if cursor was at the back
		if(cursIndx == numItems - 1){
			cursIndx = -1;
			cursor = null;
		}
		back = back.prev;
		back.next = null;
		numItems--;
	}
	// Deletes cursor element, making cursor undefined.
	// Pre: length()>0, index()>=0
	void delete(){
		if(this.isEmpty()){
			throw new RuntimeException("List ADT Error: calling delete() on an empty list");
		}
		if(cursIndx < 0){
			throw new RuntimeException("List ADT Error: calling delete() when cursor is undefined");
		}
		
		if(cursIndx == 0){
			this.deleteFront();
		}
		else if(cursIndx == numItems - 1){
			this.deleteBack();
		}
		else{
			Node temp = cursor.prev;
			temp.next = cursor.next;
			cursor.next.prev = temp;
			numItems--;
		
			cursor = null;
			cursIndx = -1;
		}
	}
	
	List copy(){
		if(this != null){
			List A = new List();
			Node temp = this.front;
			
			while(temp != null){
				A.append(temp.item);
				temp = temp.next;
			}
			return A;
		}
		return null;
	}
	// Overrides Object's toString method. Returns a String
	// representation of this List consisting of a space 
	// separated sequence of integers, with front on left.
	public String toString(){
		if(this.isEmpty()){
			return "";
		}
		else{
			String tempString = "";
			Node temp = front;
			while(temp != null){
				tempString += (String.valueOf(temp.item)).toString() + " ";
				temp = temp.next;
			}
			return tempString;
		}
	}
	
}